cf-oscore/src/test/java/org/eclipse/californium/oscore/EndpointContextInfoTest.java

```diff
@@ -0,0 +1,406 @@
+            System.out.println("Client: Request endpoint context type: " + endpointContext.getClass());
+
+            try {
+                assertNotNull(endpointContext);
+                assertEquals(MapBasedEndpointContext.class, endpointContext.getClass());
+
+                assertEquals(Boolean.TRUE.toString(), endpointContext.get(OSCORE));
+                assertEquals(sidString, endpointContext.get(OSCORE_SENDER_ID));
+                assertEquals(ridString, endpointContext.get(OSCORE_RECIPIENT_ID));
+                assertEquals(contextIdString, endpointContext.get(OSCORE_CONTEXT_ID));
+                assertEquals(ctxUri, endpointContext.get(OSCORE_URI));
+            } catch (AssertionError e) {
+                System.err.println(e.getMessage());
+                status = e.getMessage();
+                passed = false;
+            }
+
+        }
+
+        /**
+         * @return the passed
+         */
+        public boolean isPassed() {
+            return passed;
+        }
+
+        /**
+         * @return the status
+         */
+        public String getStatus() {
+            return status;
+        }
+
+    }
+
+	/**
+	 * Creates a simple server for testing.
+	 *
+	 * It will check that appropriate OSCORE information has been set
+	 * for the incoming request and the outgoing response.
+	 *
+	 * @throws OSException if creating the OSCORE context fails
+	 *
+	 */
+	private void createSimpleServer() throws OSException {
+	    //Set up OSCORE context information for response (server)
+	    byte[] sidServer = new byte[] { 0x01, 0x02, 0x03, 0x04 };
+	    byte[] ridServer = new byte[] { 0x77, 0x66, 0x55, 0x44 };
+	    final OSCoreCtx ctx = new OSCoreCtx(master_secret, true, alg, sidServer, ridServer, kdf, 32, null, context_id);
+	    String clientUri = "coap://" + TestTools.LOCALHOST_EPHEMERAL.getAddress().getHostAddress();
+	    dbServer.addContext(clientUri, ctx);
+
+	    //Create string versions of rid, sid and context ID for server
+	    final String sidServerString = DatatypeConverter.printHexBinary(sidServer);
+	    final String ridServerString = DatatypeConverter.printHexBinary(ridServer);
+	    final String contextIdString = DatatypeConverter.printHexBinary(context_id);
+
+	    final String ctxUri = ctx.getUri();
+	    assertNotNull(ctxUri);
+
+	    //Create server
+	    CoapEndpoint.Builder builder = new CoapEndpoint.Builder();
+	    builder.setCustomCoapStackArgument(dbServer);
+	    builder.setInetSocketAddress(TestTools.LOCALHOST_EPHEMERAL);
+	    serverEndpoint = builder.build();
+	    server = new CoapServer();
+	    server.addEndpoint(serverEndpoint);
+	    server.setMessageDeliverer(new MessageDeliverer() {
+	        @Override
+	        public void deliverRequest(Exchange exchange) {
+	            System.out.println("server received request");
+
+	            //Check request source context after reception
+	            assertNull(exchange.getRequest().getDestinationContext());
```

Comment:
Just to mention:
"assert" must be executed from the "main testing thread", or, at least the exception thrown by that assert must be forwarded and (re-)thrown by the "main testing thread".
Calling "assert" on a callback usually doesn't work.
For the `OSCoreObserveTest`I fixed that by introducing the `CountingCoapHandler`.
For the `MessageDeliverer` it is required to forward the exception also.

Reply:
Yes I will try to solve this in a nicer way. The way I tried to do it now was that if an assert fails  when the server is receiving a request it will not reply to the message sent by the client. So if I add assertEquals(1, 2) there the tests will fail (but with a Client received no response null pointer exception). For the MessageObserver for outgoing requests I tried using the passed and status variables.